File: apps/api/src/abf_api/routers/audit_logs.py
from fastapi import APIRouter, Depends
from supabase import Client

from abf_api.deps.supabase import get_supabase
from abf_api.models.schemas import AuditLogCreate, AuditLogRead

router = APIRouter(prefix="/audit-logs", tags=["audit-logs"])


@router.get("", response_model=list[AuditLogRead])
def list_audit_logs(
    business_id: str | None = None,
    entity_type: str | None = None,
    entity_id: str | None = None,
    limit: int = 50,
    db: Client = Depends(get_supabase),
):
    q = db.table("audit_logs").select("*").order("created_at", desc=True).limit(limit)
    if business_id:
        q = q.eq("business_id", business_id)
    if entity_type:
        q = q.eq("entity_type", entity_type)
    if entity_id:
        q = q.eq("entity_id", entity_id)
    return q.execute().data


@router.post("", response_model=AuditLogRead, status_code=201)
def create_audit_log(body: AuditLogCreate, db: Client = Depends(get_supabase)):
    data = body.model_dump(exclude_none=True)
    res = db.table("audit_logs").insert(data).execute()
    return res.data[0]

File: packages/db/src/queries/audit-logs.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type { Database, AuditLogInsert } from "../types/database";

type Client = SupabaseClient<Database>;

export function listAuditLogs(
  client: Client,
  filters?: {
    businessId?: string;
    entityType?: string;
    entityId?: string;
    limit?: number;
  }
) {
  let query = client
    .from("audit_logs")
    .select("*")
    .order("created_at", { ascending: false });

  if (filters?.businessId) {
    query = query.eq("business_id", filters.businessId);
  }
  if (filters?.entityType) {
    query = query.eq("entity_type", filters.entityType);
  }
  if (filters?.entityId) {
    query = query.eq("entity_id", filters.entityId);
  }
  if (filters?.limit) {
    query = query.limit(filters.limit);
  }

  return query;
}

export function createAuditLog(client: Client, data: AuditLogInsert) {
  return client.from("audit_logs").insert(data).select().single();
}
